### PR TITLE
chore: validateAndCastValueData

### PR DESCRIPTION
**Pull request by makeavish**

### CODE REVIEW

**Comment by makeavish** [comment]
Have to add more contextual error log

File: pkg/query-service/utils/format.go [hunk 1]
```diff
@@ -3,11 +3,72 @@ package utils
...
+			return v, nil
+		case string:
+			float64val, err:= strconv.ParseFloat(x, 64)
+			if err != nil {
+				return nil, fmt.Errorf("invalid data type, expected float, got %v", reflect.TypeOf(v))
+			}
+			return float64val, nil
+		case int:
+			return float64(x), nil
+		case int64:
```
Comment (nityanandagohain) [inline pkg/query-service/utils/format.go:62]: can be a single line, int, int32, int64

**Review by nityanandagohain — CHANGES_REQUESTED** [review]
can you make the change

File: pkg/query-service/utils/format.go (hunk 1 above)
Comment (makeavish) [reply pkg/query-service/utils/format.go:62]: Then float64(x) is not working as x is treated a interface in that case

**Review by nityanandagohain — COMMENTED** [review]
There is one edge case, ex- 
If data type is `float` and you are using the `in` operator then the value will be `[]interface{}{1, 2, 3, 4}`

right now it's returning this error `invalid data type, expected float, got []interface {}`

**Comment by makeavish** [comment]
@nityanandagohain Thanks for pointing out, resolved it

File: pkg/query-service/utils/format_test.go [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+			v:        true,
+			dataType: v3.AttributeKeyDataTypeString,
+		},
+		want:    "true",
+		wantErr: false,
+	},
+	{
+		name: "v3.AttributeKeyDataTypeString: Valid []interface{}",
+		args: args{
+			v:        []interface{}{"test"},
```
Comment (srikanthccv) [inline pkg/query-service/utils/format_test.go:70]: What happens if this is an array of numbers?

File: pkg/query-service/utils/format_test.go [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+			dataType: v3.AttributeKeyDataTypeString,
+		},
+		want:    []interface{}{"test"},
+		wantErr: false,
+	},
+	{
+		name: "v3.AttributeKeyDataTypeString: Invalid type",
+		args: args{
+			v:        map[string]interface{}{"test": "test"},
```
Comment (srikanthccv) [inline pkg/query-service/utils/format_test.go:79]: And similarly, what happens if this is a mix of numbers and strings in JSON?

Comment (makeavish) [reply pkg/query-service/utils/format_test.go:79]: We are not supporting `map[string]interface{}{"test": "test"}` so it should return error

File: pkg/query-service/utils/format_test.go (hunk 2 above)
Comment (makeavish) [reply pkg/query-service/utils/format_test.go:70]: We are not validating the interface elements as it won't cause failure at Clickhouse. Even if elements are mix of numbers and strings

**Review by nityanandagohain — APPROVED** [review]

**Review by srikanthccv — APPROVED** [review]
The key needs to be included in the error message. I am assuming that will be done in the caller function of this.

**Comment by makeavish** [comment]
Yes, the caller function would be responsible for including key in the error message.
cc: @nityanandagohain

**Pull request merged**